File: pocs/metrics-report-skill/sample-app/java-react/backend/src/main/java/com/retirement/controller/RetirementController.java
package com.retirement.controller;

import com.retirement.model.RetirementInput;
import com.retirement.model.RetirementResult;
import com.retirement.service.RetirementCalculationService;
import jakarta.validation.Valid;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.Map;

@RestController
@RequestMapping("/api/retirement")
public class RetirementController {

    private final RetirementCalculationService calculationService;

    public RetirementController(RetirementCalculationService calculationService) {
        this.calculationService = calculationService;
    }

    @PostMapping("/calculate")
    public ResponseEntity<?> calculate(@Valid @RequestBody RetirementInput input) {
        try {
            RetirementResult result = calculationService.calculate(input);
            return ResponseEntity.ok(result);
        } catch (IllegalArgumentException e) {
            return ResponseEntity.badRequest().body(Map.of("error", e.getMessage()));
        }
    }

    @GetMapping("/health")
    public ResponseEntity<Map<String, String>> health() {
        return ResponseEntity.ok(Map.of("status", "UP"));
    }
}

File: pocs/metrics-report-skill/sample-app/scala3-react/backend/src/main/scala/com/retirement/controller/RetirementController.scala
package com.retirement.controller

import com.retirement.model.{RetirementInput, RetirementResult}
import com.retirement.service.RetirementCalculationService
import jakarta.validation.Valid
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.*

@RestController
@RequestMapping(Array("/api/retirement"))
class RetirementController(calculationService: RetirementCalculationService):

  @PostMapping(Array("/calculate"))
  def calculate(@Valid @RequestBody input: RetirementInput): ResponseEntity[?] =
    try
      val result = calculationService.calculate(input)
      ResponseEntity.ok(result)
    catch
      case e: IllegalArgumentException =>
        ResponseEntity.badRequest().body(java.util.Map.of("error", e.getMessage))

  @GetMapping(Array("/health"))
  def health(): ResponseEntity[java.util.Map[String, String]] =
    ResponseEntity.ok(java.util.Map.of("status", "UP"))
